### PR TITLE
Incorporate unsubscribe and unsubscribeAll to EventEngine

### PR DESCRIPTION
**Pull request by marcin-cebo**

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt [hunk 1]
```diff
@@ -31,7 +35,8 @@ import java.util.concurrent.Executors
 import java.util.concurrent.ScheduledExecutorService
 
 class Subscribe(
-    private val eventEngineManager: EventEngineManager
+    private val eventEngineManager: EventEngineManager,
+    private val stateManager: SubscriptionState = SubscriptionState()
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt:39]: Name `stateManager` should be replaced with `subscriptionState`

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt [hunk 2]
```diff
@@ -99,36 +104,47 @@ class Subscribe(
         withPresence: Boolean,
         withTimetoken: Long = 0L,
     ) {
+        throwExceptionIfChannelAndChannelGroupIsMissing(channels, channelGroups)
+        addChannelsToLocalStorage(channels)
+        addChannelGroupsToLocalStorage(channelGroups)
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt:109]: This method (just like all other methods here) is not synchronized or do not guard the internal state from being modified concurrently and I would expect to see a lot of `ConcurrentModificationExceptions`. As already mentioned in other comment maybe we should think through where exactly should we handle that

**Review by kleewho — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/pubnub/api/managers/SubscriptionState.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.pubnub.api.managers
+
+import com.pubnub.api.models.SubscriptionItem
+
+class SubscriptionState {
+
+    internal var channels: HashMap<String, SubscriptionItem> = hashMapOf()
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/managers/SubscriptionState.kt]: ok. I will change it.

File: src/main/kotlin/com/pubnub/api/managers/SubscriptionState.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        @Synchronized internal set
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/managers/SubscriptionState.kt]: ok. I will change it.

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt (hunk 2 above)
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt]: ok. I will change it.

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt [hunk 5]
```diff
@@ -3,11 +3,14 @@
...
 
 import com.pubnub.api.PubNub
+import com.pubnub.api.PubNubError
+import com.pubnub.api.PubNubException
 import com.pubnub.api.eventengine.EffectDispatcher
 import com.pubnub.api.eventengine.EventEngineConf
 import com.pubnub.api.eventengine.Sink
 import com.pubnub.api.managers.EventEngineManager
 import com.pubnub.api.managers.ListenerManager
+import com.pubnub.api.managers.SubscriptionData
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt:13]: Can we keep it in one package?

File: src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+        assertEquals(
+            Event.SubscriptionChanged(
+                (channelsInSubscriptionData + channelToSubscribe).toList(),
+                (channelGroupsInSubscriptionData + channelGroupsToSubscribe).toList()
+            ),
+            event.captured
+        )
+        assertTrue(subscriptionData.channels.contains(CHANNEL_01))
+        assertTrue(subscriptionData.channelGroups.contains(CHANNEL_GROUPS_01))
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:58]: What does it actually test here?

Comment (kleewho) [reply src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:58]: Maybe it's enough to verify that the we have correct event? I mean, we had to get some of the information from somewhere, so if we're able to create an event with correct content then I think we can omit asserting internals

**Review by kleewho — APPROVED** [review]

File: src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    @Test
+    internal fun `should remove channels and channelGroups from local storage and pass SubscriptionChange event for handling when at least one channel or channelGroup left in storage`() {
+        val channelsToUnsubscribe: List<String> = listOf(CHANNEL_01)
+        val channelGroupsToUnsubscribe: List<String> = listOf(CHANNEL_GROUPS_01)
+
+        objectUnderTest.unsubscribe(channelsToUnsubscribe, channelGroupsToUnsubscribe)
+
+        verify { eventEngineManager.addEventToQueue(any()) }
+        assertEquals(Event.SubscriptionChanged(listOf(CHANNEL_02), listOf(CHANNEL_GROUPS_02)), event.captured)
+        assertTrue(subscriptionData.channels.size == 1 && subscriptionData.channels.contains(CHANNEL_02))
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:95]: If you really have to verify internals there are better ways to write this assertion. 

```
assertEquals(setOf(CHANNEL_02), subscriptionData.channels)
```

File: src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt (hunk 6 above)
Comment (marcin-cebo) [reply src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:58]: ok. I will modify it.

File: src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt (hunk 7 above)
Comment (marcin-cebo) [reply src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:95]: ok. I will modify it.

**Pull request merged**